include/ensmallen_bits/nsga2/nsga2_impl.hpp

```diff
@@ -0,0 +1,389 @@
+    std::vector<arma::Col<double> >& calculatedObjectives)
+{
+  for (size_t i = 0; i < populationSize; i++)
+  {
+    calculatedObjectives[i](I) = std::get<I>(objectives).Evaluate(population[i]);
+    EvaluateObjectives<I+1, MatType, ArbitraryFunctionType...>(population, objectives,
+                                                               calculatedObjectives);
+  }
+}
+
+//! Reproduce and generate new candidates.
+template<typename MatType>
+inline void NSGA2::BinaryTournamentSelection(std::vector<MatType>& population,
+                                             const arma::vec& lowerBound,
+                                             const arma::vec& upperBound)
+{
+  std::vector<MatType> children;
+
+  while (children.size() < population.size())
+  {
+    // Choose two random parents for reproduction from the elite population.
+    size_t indexA = arma::randi<size_t>(arma::distr_param(0, populationSize - 1));
+    size_t indexB = arma::randi<size_t>(arma::distr_param(0, populationSize - 1));
+
+    // Make sure that the parents differ.
+    if (indexA == indexB)
+    {
+      if (indexB < populationSize - 1)
+        indexB++;
+      else
+        indexB--;
+    }
+
+    // Initialize the children to the respective parents.
+    MatType childA = population[indexA], childB = population[indexB];
+
+    Crossover(childA, childB, population[indexA], population[indexB]);
+
+    Mutate(childA, lowerBound, upperBound);
+    Mutate(childB, lowerBound, upperBound);
+
+    // Add the children to the candidate population.
+    children.push_back(childA);
+    children.push_back(childB);
+  }
+
+  // Add the candidates to the elite population.
+  population.insert(std::end(population), std::begin(children), std::end(children));
+}
+
+//! Perform crossover of genes for the children.
+template<typename MatType>
+inline void NSGA2::Crossover(MatType& childA,
+                             MatType& childB,
+                             const MatType& parentA,
+                             const MatType& parentB)
+{
+  // Indices at which crossover is to occur.
+  const arma::umat idx = arma::randu<MatType>(childA.n_rows, childA.n_cols) < crossoverProb;
+
+  // Use traits from parentA for indices where idx is 1 and parentB otherwise.
+  childA = parentA % idx + parentB % (1 - idx);
+  // Use traits from parentB for indices where idx is 1 and parentA otherwise.
+  childB = parentA % (1 - idx) + parentA % idx;
+}
+
+//! Perform mutation of the candidates weights with some noise.
+template<typename MatType>
+inline void NSGA2::Mutate(MatType& child,
+                          const arma::vec& lowerBound,
+                          const arma::vec& upperBound)
+{
+  child += (arma::randu<MatType>(child.n_rows, child.n_cols) < mutationProb) %
+      (mutationStrength * arma::randn<MatType>(child.n_rows, child.n_cols));
+
+  // Constrain all genes to be between bounds.
+  for (size_t idx = 0; idx < numObjectives; idx++)
```

Comment:
I am trying to implement MOEA/D using this PR as a reference and I have a doubt here, is `idx < numObjectives` correct? It is not necessary that the dimensions of the variable space and objective space are same. Sorry if I am missing something obvious.

Reply:
Here, each gene of the child corresponds to a single objective.

Reply:
So an element of the population is not the coordinates of the point, but values of the objectives at that point?

Reply:
Erm no, let me rephrase, the child contains the coordinates but there is a 1:1 correspondence to the coordinates and the objectives. You, use each of the coordinates to evaluate the objectives for each of the children.

Reply:
![c62e08ff7fa8999da75995c45b4d816a](https://user-images.githubusercontent.com/64355352/87437382-dd6bf080-c60b-11ea-898c-9e2758b7f03e.jpg)
I think there is some miscommunication here. By "coordinates of the point", I mean "x belonging to S", i.e. the variable space, but you mean "F", i.e. the objective space. Is this right?

Reply:
Wait. Each child has n (`numObjectives` here) number of m-vectors, each corresponding to an objective. Each of these m dimensions correspond to variable that each of the objectives require. For example, the [Fonseca-Flemming](https://github.com/say4n/ensmallen/blob/75ae7efbc4f1ce0e99164d576d5fa595cc6ab3a4/include/ensmallen_bits/problems/fonseca_flemming_function.hpp#L42) function has 2 objectives, each requiring 3 variables.

Reply:
Ahh! I get it now. Thank you for patiently answering my doubts. Really appreciated!

Reply:
Not a problem, always a pleasure! 😄

Reply:
Hi @say4n sorry to bother you again. I thought I understood what you said but I am stuck again. I added lines to print the size of `child` and `lowerBound` just before this loop and the output for fonseca and flemming function was,
```
Size of child 3x1
Size of lower bound, should be same as child 2x1
``` 
X4000
While the output makes complete sense to me, I fail to understand how is comparing vectors of different sizes right? If number of objectives is greater than the number of variables, there will be an `indexOutOfBounds` exception and if it is less how will we ensure that the later indices of child are within bounds?
Sorry again if I am missing something obvious.

Reply:
Ah, so the thing is you only have bounds stored per objective (to avoid duplication). If you n-objectives, you will have n corresponding upper and lower bounds. You can see the definitions [here](https://github.com/say4n/ensmallen/blob/3c6aa81ca4a234bac8b49ff4ce595d3cf0e55282/tests/nsga2_test.cpp#L39) and [here](https://github.com/say4n/ensmallen/blob/3c6aa81ca4a234bac8b49ff4ce595d3cf0e55282/tests/nsga2_test.cpp#L81) for Schaffer N.1 and Fonseca Flemming functions respectively.

Reply:
Sorry again, but I think I am not able to grasp what you are saying. If,

>  Each of these three variables must lie within lowerBound(i) and upperBound(i) for the i-th objective

will it not be sufficient to have one lowerBound (the biggest element in `lowerBound` array) variable and one upperBound (the smallest in `upperBound` array) variable. I am also not comfortable with the idea of comparing  `child[idx]` with `lowerBound(idx)` only, don't we need to make sure `child[idx]` is greater than all elements in `lowerBound` (or the greatest element in `lowerBound`). Also, what will happen when number of objectives is greater than the number of variables. Wouldn't having `lowerBound` and `upperBound` per variable make more sense? No matter how hard I try, I am not able to wrap my head around this concept, any help would be greatly appreciated. Thanks!

Reply:
> I am also not comfortable with the idea of comparing child[idx] with lowerBound(idx) only, don't we need to make sure child[idx] is greater than all elements in lowerBound

The author's implementation of NSGA2 clips the variable in between the bounds. Which is what has been done [here](https://github.com/say4n/ensmallen/blob/3c6aa81ca4a234bac8b49ff4ce595d3cf0e55282/include/ensmallen_bits/nsga2/nsga2_impl.hpp#L258) post mutation.

> What will happen when number of objectives is greater than the number of variables.

It doesn't matter since the bounds have been set per objective by the author. If you need them per variable, it is trivial to add it yourself. 😄

Reply:
Just to make sure that I get it, in the paper there is an example in table 5 called WATER with 5 objectives and 3 variables. The constraints on each variable is different, x1 lies between 0.01 and 0.45 and x2 and x3 lie between 0.01 and 0.10. What will be the lowerBound and upperBound arrays in this particular example? I sincerely thank you for taking time to clarify this for me.

Reply:
What paper are you talking of?

Reply:
[This](https://www.google.com/url?sa=t&source=web&rct=j&url=https://www.iitk.ac.in/kangal/Deb_NSGA-II.pdf&ved=2ahUKEwiI_J32ktTqAhWJXSsKHawzC3MQFjAPegQIARAB&usg=AOvVaw1BcJVhQKdt9xbdiQerMtR0). The paper on NSGA2.

Reply:
Ah, so per variable bounds haven't been added to this PR. :)

Reply:
Hey @UtR491, I was wrong regarding the upper and lower bounds. I do have per variable bounds in this PR itself, I had forgotten to add the bound for the third variable for Foseca Flemming. 😭